DO sync service

jgoizueta opened the pull request:
Story details: https://app.clubhouse.io/cartoteam/story/69791

Add DO sync service to manage syncs for DO subscriptions

jgoizueta commented:
Note that when a DO sync is created, both the initial data import and the synchronization records have the table name empty. The actual name assigned to the CARTO table will be set during the initial import process both in the data import and the synchronization.

In Import API the initial name can be assigned through the `table_name` parameter, but I see no use in that, as it is not final and will be replaced eventually by the actual name.

rafatower commented:
In case it helps, here the test that fails (it took me a while to locate it):
```
Failures:

  1) Carto::Api::Public::DataObservatoryController subscriptions returns 200 with the non expired subscriptions
     Failure/Error: expect(datasets.first).to eq expected_dataset
       
       expected: {:project=>"carto", :dataset=>"abc", :table=>"table2", :id=>"carto.abc.table2", :type=>"dataset"}
            got: {:project=>"carto", :dataset=>"abc", :table=>"table2", :id=>"carto.abc.table2", :type=>"dataset", :expires_at=>"2021-06-17T18:16:39.258+00:00"}
       
       (compared using ==)
       
       Diff:
       @@ -1,4 +1,5 @@
        :dataset => "abc",
       +:expires_at => "2021-06-17T18:16:39.258+00:00",
        :id => "carto.abc.table2",
        :project => "carto",
        :table => "table2",
     # ./spec/requests/carto/api/public/data_observatory_controller_spec.rb:149:in `block (4 levels) in <top (required)>'
     # ./spec/support/helpers.rb:41:in `get_json'
     # ./spec/requests/carto/api/public/data_observatory_controller_spec.rb:145:in `block (3 levels) in <top (required)>'

Finished in 2 minutes 59.98 seconds
159 examples, 1 failure

Failed examples:

rspec ./spec/requests/carto/api/public/data_observatory_controller_spec.rb:143 # Carto::Api::Public::DataObservatoryController subscriptions returns 200 with the non expired subscriptions
```

rafatower commented on app/services/carto/do_sync_service.rb:
```diff
@@ -0,0 +1,113 @@
+      member.store
+    end
+
+    def temptative_table_name(subscription_id)
```
nitpicking: I think the correct spelling is "tentative"

jgoizueta commented:
I've rebased this to master and removed dependencies so this can be deployed first to unblock #15718

The intent is to provide a functional `DoSyncService#sync`. Changes will be made later that require integration with #15703 to discriminate invalid subscriptions etc.

Regarding the former failing test:
> PS: I guess the failure comes from the base branch: #15703
The cause is there indeed, and it's now removed from this PR.

rafatower commented on spec/services/carto/do_sync_service_spec.rb:
```diff
@@ -0,0 +1,264 @@
+      sync['sync_status'].should eq 'syncing'
+    end
+
+    # it 'does nothing for expired subscriptions' do
```
Minor: I'd remove commented out code

jgoizueta commented on app/services/carto/do_sync_service.rb:
```diff
@@ -0,0 +1,155 @@
+    end
+
+    # stop sync'ing a subscription
+    def remove_sync(subscription_id)
```
I agree but I'll leave it for the next round of changes; `create_sync` will probably raise exceptions too by then and warrant a name change too. Neither `create_sync` nor `remove_sync` are not going to be used by now and I leave them in this PR only because they are handy for tests.
